plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcConnectionCreationTest.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.jdbc;
+
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.testing.AbstractTestQueryFramework;
+import org.intellij.lang.annotations.Language;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static io.trino.testing.DataProviders.toDataProvider;
+import static java.util.Collections.synchronizedMap;
+import static java.util.Objects.requireNonNull;
+import static org.assertj.core.api.Assertions.assertThat;
+
+@Test(singleThreaded = true) // this class is stateful, see fields
+public abstract class BaseJdbcConnectionCreationTest
+        extends AbstractTestQueryFramework
+{
+    protected ConnectionCountingConnectionFactory connectionFactory;
+
+    @BeforeClass
+    public void verifySetup()
+    {
+        requireNonNull(connectionFactory, "connectionFactory is null");
+        connectionFactory.assertThatNoConnectionHasLeaked();
+    }
+
+    @Test(dataProvider = "testCases")
+    public void test(TestCase testCase)
+    {
+        int before = connectionFactory.openConnections.get();
+        if (testCase.errorMessage.isPresent()) {
+            assertQueryFails(testCase.query, testCase.errorMessage.get());
+        }
+        else {
+            getQueryRunner().execute(testCase.query);
+        }
+        int after = connectionFactory.openConnections.get();
+        assertThat(after - before).isEqualTo(testCase.expectedJdbcConnectionsCount);
+        connectionFactory.assertThatNoConnectionHasLeaked();
+    }
+
+    @DataProvider
+    public final Object[][] testCases()
+    {
+        return getTestCases().stream()
+                .collect(toDataProvider());
+    }
+
+    protected abstract List<TestCase> getTestCases();
```

Comment:
Every test class can be modeled this way: parent class that ask child class to list test cases.
Yet, we typically find it more natural to just have parent class that provide testing infrastructure (setup, assertions) and child class provide a `@Test` method(s) to invoke these assertions.

I.e. instead of `TestCase` you would have a `protected` `assert....` method that takes `String query, int expectedJdbcConnectionsCount`.